mod-permissions/src/main/resources/domain/mod-permissions/features/permissions.feature

```diff
@@ -0,0 +1,397 @@
+
+  # Check that the permission has been granted.
+  Given path 'perms/permissions'
+  And param query = 'permissionName=="' + permToAdd + '"'
+  And headers commonHeaders
+  When method GET
+  Then status 200
+  And match response.permissions[0].grantedTo contains permissionsUserId
+
+  # And that the user has the permission.
+  Given path 'perms/users/', permissionsUserId, 'permissions'
+  And headers commonHeaders
+  When method GET
+  Then status 200
+  And match response.permissionNames contains permToAdd
+
+  # Remove the permission.
+  Given path 'perms/users/', permissionsUserId, 'permissions', permToAdd
+  And headers karate.merge(commonHeaders, optionsHeaders)
+  When method OPTIONS
+  Then status 204
+  And match header access-control-allow-methods contains "DELETE"
+
+  Given path 'perms/users/', permissionsUserId, 'permissions', permToAdd
+  And headers commonHeaders
+  When method DELETE
+  Then status 204
+
+  # Check that it was removed from the permission and the user permission.
+  Given path 'perms/users/', permissionsUserId, 'permissions'
+  And headers commonHeaders
+  When method GET
+  Then status 200
+  And match response.permissionNames !contains permToAdd
+
+  Given path 'perms/permissions'
+  And param query = 'permissionName=="' + permToAdd + '"'
+  And headers commonHeaders
+  When method GET
+  Then status 200
+  And match response.permissions[0].grantedTo !contains permissionsUserId
+
+  #
+  # Test operations on permissions themselves. This is something that happens in the
+  # browser under Settings > Users > Permission sets.
+  #
+
+  Scenario: Get a certain number of permissions for the tenant and validate the response
+    * def numberToGet = 10
+    Given path 'perms/permissions'
+    And headers karate.merge(commonHeaders, optionsHeaders)
+    When method OPTIONS
+    Then status 204
+    And match header access-control-allow-methods contains "GET"
+  
+    Given path 'perms/permissions'
+    And param length = numberToGet
+    And headers commonHeaders
+    When method GET
+    Then status 200
+    And match response == { totalRecords: #number, permissions: '#[numberToGet]' }
+    # Do some schema validation. Adding or removing properties to the schema will break this 
+    # but I think that is how it should be. Note: not every object seems to have a metadata field
+    # so I'm marking that as optional (that's what the double hashtag means in karate).
+    And match each response.permissions ==
+    """
+    {
+      "id": "#uuid",
+      "permissionName": "#string",
+      "displayName": "#string",
+      "description": "#string",
+      "moduleName": "#string",
+      "moduleVersion": "#string",
+      "tags": "#array",
+      "childOf": "#array",
+      "grantedTo": "#array",
+      "subPermissions": "#array",
+      "dummy": "#boolean",
+      "mutable": "#boolean",
+      "visible": "#boolean",
+      "deprecated": "#boolean",
+      "metadata": "##object"
+    }
+    """
```

Comment:
We don't want to update the integration test when we add new properties to the permission record. As long as other modules can successfully handle the new property that integration test should succeed.

Reply:
I could be persuaded either way. If we do no schema validation then we won't catch (however unlikely) situations where an important property stops getting included. How about if we define a few properties as critical and just check for those? Good candidates might be `id`, `childOf`, `grantedTo`, `subPermissions`. 

I think the karate way of doing that would be a little more verbose but still useful:

```
And match response[0].childOf == #array
```

There probably would be a way to do it in the context of the `each` which seems like a useful construct.

Reply:
` And match each response contains { "property: "#string" }` seems like the way to validate parts of a schema using `each`.

Reply:
To catch that module A stops including an important property we write an integration test where module B fails when that property is missing. Module B can be a back-end or front-end module.

The test whether module A returns property X is a unit test, not an integration test.

Reply:
Thanks for clarifying. I like the approach. Made the changes you recommended.